### PR TITLE
PE-42104 - Adding Jobstat to SAT CLI

### PR DESCRIPTION
**Pull request by edzaka-hpe**
## Summary and Scope

New feature named "Jobstat," which allows users and admins to get a system-wide view of application status data via the SAT toolkit.  The change is backwards compatible as long as it has access to the elasticsearch api. The feature is meant to be implemented on any system running PALS. Currently, it is being developed for Slice.

## Issues and Related PRs
PE-42104 - Adding Jobstat to SAT CLI

## Testing
Feature was tested in a pyenv environment. Output of the command gave desired table to the user. 

### Tested on:
  * Local development environment
  * Virtual Shasta

### Test description:

Test included running `sat jobstat` in a virtual environment. The script sent a HTTP GET request to a flask server (which will soon be held in Slice in a Kubernetes pod running on PBSPro). The flask server returns a JSON object from the elasticsearch api via a python script on the pod. `sat jobstat` returns proper table output. Exceptions were tested and handled properly by the script as well.

## Risks and Mitigations
A risk is the elasticsearch api not working with the command.

## Pull Request Checklist

- [x ] Version number(s) incremented, if applicable
- [ x] Copyrights updated
- [ x] License file intact
- [ x] Target branch correct
- [x ] CHANGELOG.md updated
- [x ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: docs/man/sat-jobstat.8.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
+=============
+ SAT-JOBSTAT
+=============
+
+-------------------------
+Jobstat Short Description
```
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:6]: Reword this to provide a short description of jobstat

File: docs/man/sat-jobstat.8.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+**-a, --all**
+        Show all application information that is currently documented on the system.
+
+Summarize Remaining Options
+
+EXAMPLES
+========
+
+An example usage of the command:
```
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:39]: I would like to reword this so that it describes what the command does, i.e. "display a report of all jobs on the system"

File: sat/apiclient/elastic.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#
```
Comment (ekamin-hpe) [inline sat/apiclient/elastic.py:1]: This whole file can probably just be removed, I think it was initially added in my "new-command" template. At the time I was thinking we would use an API Client class to connect to elastic

File: sat/cli/jobstat/main.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+The main entry point for the jobstat subcommand.
+"""
+import logging
+import requests
+from sat.cli.jobstat import parser
+from sat.report import Report
+from sat.config import get_config_value
+
+
+URL = "http://localhost:1999/all"
```
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:33]: This is vaguely concerning to me; do you typically just run the service on the same local host as you expect to be running `sat`? I know it's a lot more work but it would be nice to define the service as a service that can run through the API gateway instead.

File: sat/cli/jobstat/main.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    return data
+
+
+def do_jobstat(args):
+    """Run the jobstat command with the given arguments.
+    Args:
+        args: The argparse.Namespace object containing the parsed arguments
+            passed to this subcommand.
+    """
+    # print('Hello jobstat')
```
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:55]: I'd just delete this line altogether

File: docs/man/sat-jobstat.8.rst [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+These options must be specified after the subcommand.
+
+**-h, --help**
+        Print the help message for 'sat jobstat'.
+
+**-a, --all**
+        Show all application information that is currently documented on the system.
+
+Summarize Remaining Options
```
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:34]: This line can be removed - it was just a placeholder originally.

File: docs/man/sat-jobstat.8.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
```
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:35]: Under the options specific to jobstat, add:
```
.. include:: _sat-format-opts.rst
.. include:: _sat-filter-opts.rst
```

This will automatically pull in the documentation for the common filter and format options that are being included in `parser.py`.

File: docs/man/sat-jobstat.8.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+| 7677b7d3-ec35-4e23-8678-92b823347939  | 49.pbs-host  | 5827  | mpiexec --ppn=1 -n2 ./signals       | completed | 2         | nid000001,nid000002 |
+| 76b783af-5c7e-4462-97c0-3dfd0f04b9b3  | 51.pbs-host  | 5827  | mpiexec --ppn=1 -n2 ./signals       | completed | 2         | nid000001,nid000002 |
+| 7a85c000-05a9-40cd-8551-00092cead801  | 48.pbs-host  | 5827  | mpiexec -n2 ./signals               | completed | 1         | nid000001           |
+| 8b5b14b3-8215-4cbe-bba4-7d7c82d48fef  | 50.pbs-host  | 5827  | mpiexec --ppn=1 -n2 ./signals       | completed | 2         | nid000001,nid000002 |
+| 957bbd4c-de3c-424d-99d6-fd2b728b2941  | 54.pbs-host  | 5827  | mpiexec --ppn=1 -n2 ./signals       | completed | 2         | nid000001,nid000002 |
+| ba977b71-40b6-40db-aa20-a583edc18be3  | 59.pbs-host  | 5827  | mpiexec --ppn=1 -n2 ./signals-sleep | completed | 2         | nid000001,nid000002 |
+| fea2726a-c531-4acd-b158-3fae197999dd  | 57.pbs-host  | 5827  | mpiexec --ppn=1 -n2 ./signals       | completed | 2         | nid000001,nid000002 |
++---------------------------------------+--------------+-------+-------------------------------------+-----------+-----------+---------------------+
```
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:54]: It might be kind of difficult to come up with exact example output, but would be nice to add an example that uses the filter options, e.g. "list all jobs that have completed", which would be done with:

```
# sat jobstat --filter state=completed
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -28,7 +28,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Added
-
+- Added new command "Jobstat." Provides system-wide view of application status data.
```
Comment (ekamin-hpe) [inline CHANGELOG.md:31]: nit: lowercase `jobstat` and I'd wrap it in backticks here instead of quotes.

File: sat/cli/jobstat/main.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    try:
+        if args.all is True:
+            server_data = unload_json(URL)
+            application_data = server_data.get("jobstat")
+            keys = []
+            for i in application_data[0]:
+                keys.append(i)
+            report = Report(
+                tuple(keys),
```
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:64]: nit: we could just use the `keys()` method here:

```suggestion
            application_data = server_data.get("jobstat")
            report = Report(
                tuple(application_data[0].keys()),
```

File: sat/cli/jobstat/main.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+                args.sort_by,
+                args.reverse,
+                get_config_value("format.no_headings"),
+                get_config_value("format.no_borders"),
+                filter_strs=args.filter_strs,
+                display_headings=args.fields,
+                print_format=args.format,
+            )
+            for i in application_data:
+                report.add_row(i)
```
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:75]: nit: slightly prefer a more meaningful variable name:
```suggestion
            for row in application_data:
                report.add_row(row)
```

File: sat/cli/jobstat/main.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+from sat.config import get_config_value
+
+
+URL = "http://localhost:1999/all"
+LOGGER = logging.getLogger(__name__)
+
+
+def unload_json(url):
+    """Unloads the JSON object from the server url and formats the data as a dictionary"""
+    data = {}
```
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:39]: nit: unused variable

File: sat/cli/jobstat/main.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+"""
+The main entry point for the jobstat subcommand.
+"""
+import logging
+import requests
+from sat.cli.jobstat import parser
```
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:28]: nit: unused import

**Review by ekamin-hpe — CHANGES_REQUESTED** [review]
I have a lot of small comments that it would be nice to address here. Overall the only one that I think needs some further discussion is the expectation that the service is running on localhost. Other than that this seems pretty good.

File: sat/cli/jobstat/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vinaybabug-hpe) [reply sat/cli/jobstat/main.py:33]: jobstat is running inside a pod. So, changed it to https://api-gw-service-nmn.local/apis/statechecker/jobstat/all

[the remark this answers is not in the document]
Comment (ekamin-hpe) [reply sat/cli/jobstat/main.py:33]: Sounds good, that is much closer to what we would want here. SAT has a configuration file that lets you configure the API gateway hostname, so I would suggest we use that instead of hard-coding `api-gw-service-nmn.local` here.

E.g. something like:
```
URL = f'https://{get_config_value("api_gateway.host")/apis/statechecker/jobstat/all'
```

Follow-up question, is there or will there be any authentication on this endpoint? If so (and possibly even if not) it might make sense to create a subclass of `APIGatewayClient` ([source](https://github.com/Cray-HPE/sat/blob/integration/sat/apiclient/gateway.py#L69)) - and put some of the query logic into this class instead of in the `main` function. I think this would definitely make unit testing easier.

Then it would be a little more consistent with some of the other commands. Here are some examples of these type of classes:
* BOS: https://github.com/Cray-HPE/sat/blob/integration/sat/apiclient/bos.py
* CAPMC: https://github.com/Cray-HPE/sat/blob/integration/sat/apiclient/capmc.py
* CFS: https://github.com/Cray-HPE/sat/blob/integration/sat/apiclient/cfs.py

Doing that would end up removing the need to define `URL` at all above, it would also be a little bit of a refactor of the current implementation.

**Comment by vinaybabug-hpe** [comment]
Please check PR-58 for updates on this.

**Comment by ekamin-hpe** [comment]
This is duplicated by #58

File: docs/man/sat-jobstat.8.rst (hunk 8 above)
Comment (vinaybabug-hpe) [reply docs/man/sat-jobstat.8.rst:54]: This is a request for a new command option.

Comment (ekamin-hpe) [reply docs/man/sat-jobstat.8.rst:54]: You may have not realized it but it already exists because Elmir added the format and filter options in `add_jobstat_subparser` (see the function on line 30 of `sat/cli/jobstat/parser.py`:

```
def add_jobstat_subparser(subparsers):
    """Add the jobstat parser to the parent parser.
    Args:
        subparsers: The argparse.ArgumentParser object returned by the
            add_subparsers method.
    Returns:
        None
    """
    format_options = sat.parsergroups.create_format_options()
    filter_options = sat.parsergroups.create_filter_options()
```

It hooks into the `Report` class that is being used in `main.py`, so this should already work. Is it possible to test it to verify?

File: docs/man/sat-jobstat.8.rst [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+SYNOPSIS
+========
+
+**sat** [global-opts] **jobstat** [options]
+
+DESCRIPTION
+===========
+
+Jobstat is a command that allows a user or admin to access application and job data via the command line. The command provides a table showing the user the apid, jobid, user, state, and time-reported of all jobs on the system.
```
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:21]: Sorry for not noticing this earlier, but could this line be wrapped at 80 characters?

File: docs/man/sat-jobstat.8.rst (hunk 8 above)
Comment (vinaybabug-hpe) [reply docs/man/sat-jobstat.8.rst:54]: I have not gotten to test this yet. Elmir's gave a demo, so know that code works. But I will do a round of testing after addressing review comments.

Comment (vinaybabug-hpe) [reply docs/man/sat-jobstat.8.rst:54]: I will make the current example to have the filter and give a more general example later.

Comment (ekamin-hpe) [reply docs/man/sat-jobstat.8.rst:54]: Testing would be very helpful. In the meantime I can confirm that the format and filter options are in the help text:
```
mbp-ekamin:sat ekamin$ sat jobstat --help
usage: sat jobstat [-h] [--format {pretty,yaml,json}] [--no-borders]
                   [--no-headings] [--reverse] [--sort-by FIELD]
                   [--show-empty] [--show-missing] [--fields FIELDS]
                   [--filter QUERY] [--all]

This command shows all applications and relevent information pertaining to
those applications.

optional arguments:
  -h, --help            show this help message and exit
  --all, -a             Show all application information available on the
                        system.

format options:
  Options to modify output formatting.

  --format {pretty,yaml,json}
                        Display information in the given format. Defaults to
                        'pretty'.
  --no-borders          Omit borders from tables.
  --no-headings         Omit headings from tables.
  --reverse             Sort the output in reverse order. Only applies for
                        "pretty".
  --sort-by FIELD       Select which column to sort by. Can accept a column
                        name or a 0-based index. Enclose the column name in
                        double quotes if it contains a space.
  --show-empty          Show values for columns even if every value is EMPTY.
  --show-missing        Show values for columns even if every value is MISSING
  --fields FIELDS       Display only the given comma-separated list of fields.

filtering options:
  Options to filter output.

  --filter QUERY        Filter rows of the output based on the query provided.
                        Refer to the man page for this subcommand for more
                        details regarding filter query syntax.
```

Regarding testing, I am particularly unsure about whether the URL change works or if you will need additional changes to handle API authentication as I mention in this comment: https://github.com/Cray-HPE/sat/pull/38#discussion_r968926035

File: sat/cli/jobstat/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vinaybabug-hpe) [reply sat/cli/jobstat/main.py:33]: Replaced this with code
`        connect = APIGatewayClient()
        return connect.get(['statechecker', 'jobstat', 'all']).json()`
